ostd/src/sync/rcu/non_null.rs

```diff
@@ -0,0 +1,241 @@
+        // SAFETY: The reference is created through `NonNullPtr::raw_as_ref`, hence
+        // the original owned pointer and target must outlive the lifetime parameter `'a`,
+        // and during `'a` no mutable references to the pointer will exist.
+        unsafe { &*(self.deref().deref() as *const T) }
+    }
+}
+
+unsafe impl<T: Send + Sync + 'static> NonNullPtr for Arc<T> {
+    type Target = T;
+
+    type Ref<'a>
+        = ArcRef<'a, T>
+    where
+        Self: 'a;
+
+    fn into_raw(self) -> NonNull<Self::Target> {
+        let ptr = Arc::into_raw(self).cast_mut().cast();
+
+        // SAFETY: The pointer representing an `Arc` can never be NULL.
+        unsafe { NonNull::new_unchecked(ptr) }
+    }
+
+    unsafe fn from_raw(ptr: NonNull<Self::Target>) -> Self {
+        let ptr = ptr.as_ptr().cast_const().cast();
+
+        // SAFETY: The safety is upheld by the caller.
+        unsafe { Arc::from_raw(ptr) }
+    }
+
+    unsafe fn raw_as_ref<'a>(raw: NonNull<Self::Target>) -> Self::Ref<'a> {
+        // SAFETY: By the safety requirements of `NonNullPtr::raw_as_ref`, the original pointer
+        // outlives the lifetime parameter `'a` and during `'a` no mutable references to it can
+        // exist. Thus, a shared reference to the original pointer can be created.
+        unsafe {
+            ArcRef {
+                inner: ManuallyDrop::new(Arc::from_raw(raw.as_ptr().cast())),
+                _marker: PhantomData,
+            }
+        }
+    }
+
+    fn ref_as_raw(ptr_ref: Self::Ref<'_>) -> NonNull<Self::Target> {
+        NonNullPtr::into_raw(ManuallyDrop::into_inner(ptr_ref.inner))
+    }
+}
+
+/// A type that represents `&'a Weak<T>`.
+#[derive(Debug)]
+pub struct WeakRef<'a, T: Send + Sync + 'static> {
+    inner: ManuallyDrop<Weak<T>>,
+    _marker: PhantomData<&'a Weak<T>>,
+}
+
+impl<T: Send + Sync + 'static> Deref for WeakRef<'_, T> {
+    type Target = Weak<T>;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+unsafe impl<T: Send + Sync + 'static> NonNullPtr for Weak<T> {
+    type Target = T;
+
+    type Ref<'a>
+        = WeakRef<'a, T>
+    where
+        Self: 'a;
+
+    fn into_raw(self) -> NonNull<Self::Target> {
+        let ptr = Weak::into_raw(self).cast_mut().cast();
+        // SAFETY: The pointer representing an `Weak` can never be NULL.
+        unsafe { NonNull::new_unchecked(ptr) }
+    }
+
+    unsafe fn from_raw(ptr: NonNull<Self::Target>) -> Self {
+        let ptr = ptr.as_ptr().cast_const().cast();
+
+        // SAFETY: The safety is upheld by the caller.
+        unsafe { Weak::from_raw(ptr) }
+    }
+
+    unsafe fn raw_as_ref<'a>(raw: NonNull<Self::Target>) -> Self::Ref<'a> {
+        // SAFETY: By the safety requirements of `NonNullPtr::raw_as_ref`, the original pointer
+        // outlives the lifetime parameter `'a` and during `'a` no mutable references to it can
+        // exist. Thus, a shared reference to the original pointer can be created.
```

Comment:
This is just repeating the documentations of the trait method. We can simplify this to `// SAFETY: The safety is upheld by the caller`.

Same for `<Arc<T> as NonNullPtr>::raw_as_ref`.